### PR TITLE
impl(pubsub): only add links that are sampled

### PR DESCRIPTION
**Pull request by alevenberg**
It needs to increase the span catcher buffer size since the default is 100 (https://github.com/open-telemetry/opentelemetry-cpp/blob/ca08c5a34ad4af1b6b392c8bf5fc26a5210a2f2c/exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h#L17)

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13043)

### CODE REVIEW

File: google/cloud/pubsub/internal/tracing_message_batch_test.cc [hunk 1]
```diff
@@ -130,11 +130,43 @@ TEST(TracingMessageBatch, Flush) {
...
+  auto end_spans = message_batch->Flush();
+  end_spans(make_ready_future());
+
+  auto spans = span_catcher->GetSpans();
+  EXPECT_THAT(spans,
+              Contains(AllOf(SpanHasInstrumentationScope(), SpanKindIsClient(),
+                             SpanNamed("BatchSink::AsyncPublish"),
+                             SpanHasAttributes(OTelAttribute<std::int64_t>(
+                                 sc::kMessagingBatchMessageCount, 2)),
+                             SpanHasLinks(AllOf(
```
Comment (dbolduc) [inline google/cloud/pubsub/internal/tracing_message_batch_test.cc:159]: I think to verify that `unsampled_span` is not linked to, we want `SpanLinksAre`.

**Review by dbolduc — APPROVED** [review]

File: google/cloud/pubsub/internal/tracing_message_batch_test.cc (hunk 1 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/tracing_message_batch_test.cc]: done

Comment (alevenberg) [reply google/cloud/pubsub/internal/tracing_message_batch_test.cc]: I was using the kMessagingBatchMessageCount also kind of verified it....

what you suggested is much better.

Thanks!

**Comment by alevenberg** [comment]
Closes https://github.com/googleapis/google-cloud-cpp/issues/12957

**Pull request merged**